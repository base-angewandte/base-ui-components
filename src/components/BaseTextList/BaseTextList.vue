<template>
  <div class="base-text-list">
    <div
      v-for="(item, k) in data"
      :key="k"
      class="base-text-list-group">
      <component
        :is="renderLabelAs"
        class="base-text-list-label">
        {{ item.label }}
      </component>

      <!-- String as text -->
      <!-- eslint-disable -->
      <p
        v-if="typeof item.data === 'string'"
        class="base-text-list-content base-text-list-content-pre-line">{{ item.data }}</p>
      <!-- eslint-enable -->

      <!-- Array as unordered list -->
      <ul
        v-if="typeof item.data === 'object' && typeof item.data[0] === 'string'"
        class="base-text-list-content">
        <li
          v-for="(a, l) in item.data"
          :key="l">
          {{ a }}
        </li>
      </ul>

      <!-- Array/Objects as data list -->
      <dl
        v-if="typeof item.data === 'object' && typeof item.data[0] === 'object'"
        class="base-text-list-content">
        <template
          v-for="(o, l) in item.data">
          <dt :key="'l' + l">
            {{ o.label }}:
          </dt>
          <dd :key="'v' + l">
            {{ o.value }}
          </dd>
        </template>
      </dl>
    </div>
  </div>
</template>

<script>
export default {
  name: 'BaseTextList',
  props: {
    /**
     * Data to display
     */
    data: {
      type: Array,
      default: () => ([]),
    },
    /**
     * Render component as e.g.: 'h2' | 'h3'
     */
    renderLabelAs: {
      type: String,
      default: 'div',
    },
  },
};
</script>

<style lang="scss" scoped>
  @import "../../styles/variables";

  .base-text-list {

    .base-text-list-group {

      &:first-of-type {
        .base-text-list-label {
          margin-top: 0;
        }
      }

      .base-text-list-label {
        margin-top: $line-height;
      }

      @media screen and (max-width: $mobile) {
        &:first-of-type {
          margin-top: $line-height;
        }
      }
    }

    .base-text-list-content {
      color: $font-color-second;
    }

    .base-text-list-content-pre-line {
      white-space: pre-line;
    }

    dt,
    dd {
      display: inline;
    }

    dd:after {
      display: block;
      content: '';
      width: 100%;
    }
  }
</style>
