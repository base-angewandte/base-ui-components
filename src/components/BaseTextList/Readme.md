Example of BaseTextList to render data in one column

```vue
<template>
  <div>
    <BaseTextList
      render-label-as="h2"
      :data="data" />
  </div>
</template>

<script>
import BaseTextList from './BaseTextList';
export default {
  components: {
    BaseTextList,
  },
  data() {
    return {
      data: [
        {
          label: 'Vero eos et accusam',
          data: 'no sea takimata sanctus est Lorem ipsum dolor sit amet.',
        },
        {
          label: 'Vero eos et accusam',
          data: [
            'Lorem ipsum dolor sit amet',
            'Consetetur sadipscing elitr, sed diam nonumy',
            'Eirmod tempor invidunt',
          ],
        },
        {
          label: 'Vero eos et accusam',
          data: [
            {
              label: 'www',
              value: 'http://base.uni-ak.ac.at',
              url: 'http://base.uni-ak.ac.at',
            },
            {
              label: 'email',
              value: 'base@uni-ak.ac.at',
              url: 'mailto:base@uni-ak.ac.at',
            },
            {
              label: 'Labore',
              value: 'Dolore magna aliquyam erat',
            },
          ],
        },
        {
          label: 'Vero eos et accusam',
          data: [
            {
              value: 'additional info in toolip',
              additional: [
                {
                  label: 'www',
                  value: 'base',
                  url: 'https://base.uni-ak.ac.at',
                },
                {
                  label: 'label',
                  value: 'some value',
                },
              ],
            },
            {
              value: 'external link',
              url: 'https://base.uni-ak.ac.at',
            },
            {
              value: 'internal link',
              source: 'some-id',
            },
          ],
        },
      ],
    }
  }
}
</script>
```

Example of BaseTextList to render data in two columns

```vue
<template>
  <div>
    <BaseTextList
      render-label-as="h2"
      :data="data"
      :cols2="true" />
  </div>
</template>

<script>
import BaseTextList from './BaseTextList';
export default {
  components: {
    BaseTextList,
  },
  data() {
    return {
      data: [
        {
          label: 'Vero eos et accusam',
          data: 'no sea takimata sanctus est Lorem ipsum dolor sit amet.',
        },
        {
          label: 'Vero eos et accusam',
          data: [
            'Lorem ipsum dolor sit amet',
            'Consetetur sadipscing elitr, sed diam nonumy',
            'Eirmod tempor invidunt',
          ],
        },
        {
          label: 'Vero eos et accusam',
          data: [
            {
              label: 'www',
              value: 'http://base.uni-ak.ac.at',
              url: 'http://base.uni-ak.ac.at',
            },
            {
              label: 'email',
              value: 'base@uni-ak.ac.at',
              url: 'mailto:base@uni-ak.ac.at',
            },
            {
              label: 'Labore',
              value: 'Dolore magna aliquyam erat',
            },
          ],
        },
      ],
    }
  }
}
</script>
```

Note: If data is a single Object and field type is 'string', then data will be rendered in two columns

```vue
<template>
  <div>
    <BaseTextList
      render-label-as="h2"
      :data="data"
      :cols2="true" />
  </div>
</template>

<script>
import BaseTextList from './BaseTextList';
export default {
  components: {
    BaseTextList,
  },
  data() {
    return {
      data: [
        {
          label: 'Vero eos et accusam',
          data: `Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua.
          `,
        },
      ],
    }
  }
}
</script>
```
