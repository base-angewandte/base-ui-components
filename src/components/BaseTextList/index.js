import { use, registerComponent } from '@/utils/plugins';
import BaseTextList from './BaseTextList';

const Plugin = {
  install(Vue) {
    registerComponent(Vue, BaseTextList);
  },
};

use(Plugin);

export default Plugin;

export {
  BaseTextList,
};
