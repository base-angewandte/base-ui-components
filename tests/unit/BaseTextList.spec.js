import { shallowMount } from '@vue/test-utils';
import BaseTextList from '@/components/BaseTextList/BaseTextList';

describe('BaseTextList', () => {
  const build = () => {
    const wrapper = shallowMount(BaseTextList, {
      propsData: {
        data: [
          {
            label: 'Vero eos et accusam',
            data: 'no sea takimata sanctus est Lorem ipsum dolor sit amet.',
          },
          {
            label: 'Vero eos et accusam',
            data: [
              'Lorem ipsum dolor sit amet',
              'Consetetur sadipscing elitr, sed diam nonumy',
              'Eirmod tempor invidunt',
            ],
          },
          {
            label: 'Vero eos et accusam',
            data: [
              {
                label: 'www',
                value: 'http://base.uni-ak.ac.at',
                url: 'http://base.uni-ak.ac.at',
              },
              {
                label: 'email',
                value: 'base@uni-ak.ac.at',
                url: 'mailto:base@uni-ak.ac.at',
              },
              {
                label: 'Labore',
                value: 'Dolore magna aliquyam erat',
              },
            ],
          },
        ],
      },
    });

    return {
      wrapper,
      p: () => wrapper.find('p'),
      ul: () => wrapper.find('ul'),
      li: () => wrapper.find('li'),
      dl: () => wrapper.find('dl'),
      dt: () => wrapper.find('dt'),
      dd: () => wrapper.find('dd'),
      a: () => wrapper.find('a'),
      baseClass: () => wrapper.find('.base-text-list'),
    };
  };

  it('renders the component', () => {
    // arrange
    const {
      wrapper,
      baseClass,
    } = build();

    // assert
    expect(wrapper.html())
      .toMatchSnapshot();

    expect(baseClass()
      .exists())
      .toBe(true);
  });

  it('renders <p> if object data value is a string', () => {
    // arrange
    const { p } = build();

    // assert
    expect(p()
      .exists())
      .toBe(true);
  });

  it('renders <ul> if object data value is an array', () => {
    // arrange
    const { ul, li } = build();

    // assert
    expect(ul()
      .exists())
      .toBe(true);
    expect(li()
      .exists())
      .toBe(true);
  });

  it('renders <dl> if object data value is an object', () => {
    // arrange
    const { a, dl, dt, dd } = build();

    // assert
    expect(dl()
      .exists())
      .toBe(true);
    expect(dt()
      .exists())
      .toBe(true);
    expect(dd()
      .exists())
      .toBe(true);
    expect(a()
      .exists())
      .toBe(true);
  });
});
