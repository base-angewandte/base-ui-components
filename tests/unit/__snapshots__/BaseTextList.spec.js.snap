// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTextList renders the component 1`] = `
<div class="base-text-list">
  <div class="base-text-list-group">
    <div class="base-text-list-label">
      Vero eos et accusam
    </div>
    <p class="base-text-list-content base-text-list-content-pre-line">no sea takimata sanctus est Lorem ipsum dolor sit amet.</p>
    <!---->
    <!---->
  </div>
  <div class="base-text-list-group">
    <div class="base-text-list-label">
      Vero eos et accusam
    </div>
    <!---->
    <ul class="base-text-list-content">
      <li>
        Lorem ipsum dolor sit amet
      </li>
      <li>
        Consetetur sadipscing elitr, sed diam nonumy
      </li>
      <li>
        Eirmod tempor invidunt
      </li>
    </ul>
    <!---->
  </div>
  <div class="base-text-list-group">
    <div class="base-text-list-label">
      Vero eos et accusam
    </div>
    <!---->
    <!---->
    <dl class="base-text-list-content">
      <dt class="base-text-list__content-list-item">
        www:
      </dt>
      <dd class="base-text-list__content-list-description base-text-list__content-list-item"><a href="http://base.uni-ak.ac.at" title="http://base.uni-ak.ac.at" class="base-text-list__content-link">
          http://base.uni-ak.ac.at
        </a></dd>
      <dt class="base-text-list__content-list-item">
        email:
      </dt>
      <dd class="base-text-list__content-list-description base-text-list__content-list-item"><a href="mailto:base@uni-ak.ac.at" title="base@uni-ak.ac.at" class="base-text-list__content-link">
          base@uni-ak.ac.at
        </a></dd>
      <dt class="base-text-list__content-list-item">
        Labore:
      </dt>
      <dd class="base-text-list__content-list-description base-text-list__content-list-item">
        Dolore magna aliquyam erat
      </dd>
    </dl>
  </div>
</div>
`;
